pulp_2to3_migration/app/plugin/rpm/pulp_2to3_models.py

```diff
@@ -524,3 +536,239 @@ class Meta:
         Create a Pulp 3 Distribution content for saving later in a bulk operation.
         """
         pass
+
+
+class Pulp2PackageLangpacks(Pulp2to3Content):
+    """
+    Pulp 2to3 detail content model to store pulp 2 package_langpacks
+    content details for Pulp 3 content creation.
+    """
+    matches = JSONField()
+    repo_id = models.TextField()
+
+    pulp2_type = 'package_langpacks'
+
+    class Meta:
+        unique_together = ('repo_id', 'pulp2content')
+        default_related_name = 'package_langpacks_detail_model'
+
+    @classmethod
+    async def pre_migrate_content_detail(cls, content_batch):
+        """
+        Pre-migrate Pulp 2 langpacks content with all the fields needed to create a Pulp 3 content.
+
+        Args:
+             content_batch(list of Pulp2Content): pre-migrated generic data for Pulp 2 content.
+
+        """
+        pulp2_id_obj_map = {pulp2content.pulp2_id: pulp2content for pulp2content in content_batch}
+        pulp2_ids = pulp2_id_obj_map.keys()
+        pulp2_content_batch = pulp2_models.PackageLangpacks.objects.filter(id__in=pulp2_ids)
+        pulp2langpacks_to_save = [
+            cls(repo_id=p.repo_id,
+                matches=p.matches,
+                pulp2content=pulp2_id_obj_map[p.id])
+            for p in pulp2_content_batch]
+        cls.objects.bulk_create(pulp2langpacks_to_save, ignore_conflicts=True)
+
+    async def create_pulp3_content(self):
+        """
+        Create a Pulp 3 Package Langpacks content for saving it later in a bulk operation.
+        """
+        langpacks = langpacks_to_libcomps(self)
+        langpacks_dict = PackageLangpacks.libcomps_to_dict(langpacks)
+        return (PackageLangpacks(matches=langpacks_dict['matches'],
+                                 digest=dict_digest(langpacks_dict)), None)
+
+
+class Pulp2PackageGroup(Pulp2to3Content):
+    """
+    Pulp 2to3 detail content model to store pulp 2 package_group
+    content details for Pulp 3 content creation.
+    """
+    package_group_id = models.TextField()
+    repo_id = models.TextField()
+
+    default = models.BooleanField()
+    user_visible = models.BooleanField()
+    display_order = models.IntegerField(null=True)
+    name = models.TextField()
+    description = models.TextField()
+
+    # merge optional, conditional, default, mandatory
```

Comment:
I'm not sure what's this comment is about.

Reply:
ok, so this is an interesting one. Order of the elements in the list matters. I needed to explicitly  first have mandatory, then  default, then optional and as last conditional.
Look at this output( these are results before I properly placed types of the packages in the list).
These are 2 same Package groups, one was migrated from mongo. Second synced from a repo into pulp3. So when you calculate digest, obviously it is different..
```
In [8]: x[1].packages                                                                                                                                                                                              
Out[8]: 
[{'name': 'autoconf', 'type': 3, 'requires': None, 'basearchonly': None},
 {'name': 'automake', 'type': 3, 'requires': None, 'basearchonly': None},
 {'name': 'binutils', 'type': 3, 'requires': None, 'basearchonly': None},
 {'name': 'bison', 'type': 3, 'requires': None, 'basearchonly': None},
 {'name': 'flex', 'type': 3, 'requires': None, 'basearchonly': None},
 {'name': 'gcc', 'type': 3, 'requires': None, 'basearchonly': None},
 {'name': 'gcc-c++', 'type': 3, 'requires': None, 'basearchonly': None},
 {'name': 'gettext', 'type': 3, 'requires': None, 'basearchonly': None},
 {'name': 'libtool', 'type': 3, 'requires': None, 'basearchonly': None},
 {'name': 'make', 'type': 3, 'requires': None, 'basearchonly': None},
 {'name': 'patch', 'type': 3, 'requires': None, 'basearchonly': None},
 {'name': 'pkgconfig', 'type': 3, 'requires': None, 'basearchonly': None},
 {'name': 'redhat-rpm-config',
  'type': 3,
  'requires': None,
  'basearchonly': None},
 {'name': 'rpm-build', 'type': 3, 'requires': None, 'basearchonly': None},
 {'name': 'rpm-sign', 'type': 3, 'requires': None, 'basearchonly': None},
 {'name': 'byacc', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'cscope', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'ctags', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'diffstat', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'doxygen', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'elfutils', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'gcc-gfortran', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'git', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'indent', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'intltool', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'patchutils', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'rcs', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'subversion', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'swig', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'systemtap', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'ElectricFence', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'ant', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'babel', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'bzr', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'chrpath', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'cmake', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'compat-gcc-44', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'compat-gcc-44-c++',
  'type': 1,
  'requires': None,
  'basearchonly': None},
 {'name': 'cvs', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'dejagnu', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'expect', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'gcc-gnat', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'gcc-objc', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'gcc-objc++', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'imake', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'javapackages-tools',
  'type': 1,
  'requires': None,
  'basearchonly': None},
 {'name': 'ksc', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'mercurial', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'mod_dav_svn', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'nasm', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'perltidy', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'rpmdevtools', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'rpmlint', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'systemtap-sdt-devel',
  'type': 1,
  'requires': None,
  'basearchonly': None},
 {'name': 'systemtap-server',
  'type': 1,
  'requires': None,
  'basearchonly': None}]

In [9]: x[0].packages                                                                                                                                                                                              
Out[9]: 
[{'name': 'byacc', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'cscope', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'ctags', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'diffstat', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'doxygen', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'elfutils', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'gcc-gfortran', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'git', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'indent', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'intltool', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'patchutils', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'rcs', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'subversion', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'swig', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'systemtap', 'type': 0, 'requires': None, 'basearchonly': None},
 {'name': 'ElectricFence', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'ant', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'babel', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'bzr', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'chrpath', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'cmake', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'compat-gcc-44', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'compat-gcc-44-c++',
  'type': 1,
  'requires': None,
  'basearchonly': None},
 {'name': 'cvs', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'dejagnu', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'expect', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'gcc-gnat', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'gcc-objc', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'gcc-objc++', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'imake', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'javapackages-tools',
  'type': 1,
  'requires': None,
  'basearchonly': None},
 {'name': 'ksc', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'mercurial', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'mod_dav_svn', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'nasm', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'perltidy', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'rpmdevtools', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'rpmlint', 'type': 1, 'requires': None, 'basearchonly': None},
 {'name': 'systemtap-sdt-devel',
  'type': 1,
  'requires': None,
  'basearchonly': None},
 {'name': 'systemtap-server',
  'type': 1,
  'requires': None,
  'basearchonly': None},
 {'name': 'autoconf', 'type': 3, 'requires': None, 'basearchonly': None},
 {'name': 'automake', 'type': 3, 'requires': None, 'basearchonly': None},
 {'name': 'binutils', 'type': 3, 'requires': None, 'basearchonly': None},
 {'name': 'bison', 'type': 3, 'requires': None, 'basearchonly': None},
 {'name': 'flex', 'type': 3, 'requires': None, 'basearchonly': None},
 {'name': 'gcc', 'type': 3, 'requires': None, 'basearchonly': None},
 {'name': 'gcc-c++', 'type': 3, 'requires': None, 'basearchonly': None},
 {'name': 'gettext', 'type': 3, 'requires': None, 'basearchonly': None},
 {'name': 'libtool', 'type': 3, 'requires': None, 'basearchonly': None},
 {'name': 'make', 'type': 3, 'requires': None, 'basearchonly': None},
 {'name': 'patch', 'type': 3, 'requires': None, 'basearchonly': None},
 {'name': 'pkgconfig', 'type': 3, 'requires': None, 'basearchonly': None},
 {'name': 'redhat-rpm-config',
  'type': 3,
  'requires': None,
  'basearchonly': None},
 {'name': 'rpm-build', 'type': 3, 'requires': None, 'basearchonly': None},
 {'name': 'rpm-sign', 'type': 3, 'requires': None, 'basearchonly': None}]

```

Reply:
Yeah, I see. I got it also from your _get_packages method. 
Does it make sense to have them in correct order in the comment? or if it's already mentioned in the _get_packages, just have "packages of all types are merged", or something like that. 
If one is not aware of comps details, the comment is confusing. E.g. At first. I was not even sure if it's related to the packages, apparently there are conditional packages :/

Reply:
if someone is not familiar with comps details ""packages of all types are merged" will not help either. i would leave this as is especially if _get_packages give more explanation.
I will change it to correct order :)